### PR TITLE
Implement ICD calulations in planner output

### PR DESCRIPTION
**Pull request by willemferguson**
This patch implements the calculation and printing of icd parameters
in the dive planner output.

This pull request replaces a previous pull request that appears to have become unusable
because of stupidity on my side. The changes here reflect review of the previous PR.

Signed-off-by: Willem Ferguson <willemferguson@zoology.up.ac.za>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

1) icd paramaters are calculated following the rule-of-fifths for
   OC trimix dives. For each gas change on the ascent that involves
   helium, the change in gas composition as well as the change in
   partial pressures are shown. If rule-of-fifths is exceeded, a
   warning message is printed.
2) An independent function is provided that in principle enables
   calculation of icd paramaters outside of the context of the
   dive planner.

### Changes made:

A new html table is created after the gas useage info. This gives values for dHe, dN2 and max_dN2 using rule-of-fifths. Both gas fractions and gas partial pressures are provided.

### Mentions:

@atdotde 
@sfuchs79 
@neolit123 
@Jan Mulder

### CODE REVIEW

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: There is a open decision here: Shall we print the ICD table also for the "plan_verbatim" or not. With the current implementation it's half-half which is nok - please look at the code carefully and/or test this!
If you don't want it for the "plan_verbatim", simply put this code INSIDE the if starting at new line 139.

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: This important string "Isobaric..." is missing translation.

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: Coding style?

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: In this whole block imho it has to be always `gasmix` NOT `newgasmix`

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: Coding style?

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: Coding style: New line plus indentation before `goto`

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: Again coding style concerning the "if".

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: Again the "if".

**Review by sfuchs79 — CHANGES_REQUESTED** [review]
Much nicer than all the versions before but still a few minor items to correct.
I didn't test it yet.

**Comment by sfuchs79** [comment]
Now tested and found these additional issues:
- All the "%" in the printout are missing. To print a "%" with printf you have to put `%%` in the string.
- Please don't put a space between any value and unit (%, bar)

**Comment by dirkhh** [comment]
@sfuchs79 thank you for all the great feedback
@willemferguson I'm excited to see code from you. I want to make sure we provide you feedback in an easy to use manner - so please let us know if this is working for you...

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: Code changed so that it also prints for verbatim mode.

**Comment by willemferguson** [comment]
I have no idea what it means when travis cannot build the project on Mac, as is evidenced by the above error message. Do I need to do anything?

**Comment by janmulder** [comment]
@willemferguson there are problems at Travis, and I'm monitoring the progress, and I will kick the builds when they say they are ready. No need for you to do anything.

**Comment by willemferguson** [comment]
Since there is some delay in getting code reviewed, I take the liberty of pushing my newest proposal for changes to plannernotes.c. These include mostly cosmetic changes involving addition of comments as well as small stylistic changes. No changes to the operational aspects of the code.

**Comment by neolit123** [comment]
small typo in the commit title - `calulations`.
you can easily amend the commit message by doing this in the same local branch from where the commit originated:
```
git commit --amend
< a text editor should show up; edit the message >
git push -f <remote> <remote branch>
```

**Comment by willemferguson** [comment]
Fixed. There were more than two typos. Thanks, neolit123.

File: core/plannernotes.c
Comment (atdotde) [inline core/plannernotes.c]: Is there a reason you passion icd_data by value rather than by pointer?
You could omit passing in icdlen and rather pass in icdbuffer+icdlen and bufsize-icdlen and return the increase.

File: core/plannernotes.c
Comment (atdotde) [inline core/plannernotes.c]: There is little point doing this division here. You use the result only in places where you still do another floating point division (to convert permille to fraction).

File: core/plannernotes.c
Comment (atdotde) [inline core/plannernotes.c]: Please don't make dp an argument: This only exists for planned dives (and we could do this test also for gas changes in real logged dives) and you only use it to copy the time to the icd_data struct (where I honestly don't really understand why you need the time there anyway, you could pass that as an extra argument to add_icd_entry).

**Review by atdotde — COMMENTED** [review]
Small fixes that would make this code more useful in other places. Apart from those this is fine as it is (but I will look into moving/copying this information form the planner notes to the gas change events in the profile).

File: core/plannernotes.c
Comment (sfuchs79) [inline core/plannernotes.c]: Here it has to be `gasmix` instead of `newgasmix`

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: shouldn't we align the above comment lines `2) ... 3)...etc` to start with the same indentation as `1)`.
this should be done with spaces instead of tabs.

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: ditto - alignment `i) and ii)`.

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: the conditions can be formed as:
`return get_he(oldgasmix) && results->dN2 > 0 && 5 * results->dN2 > -results->dHe;`
without the extra braces.

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: please, see my previous notes about comment alignment.

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: i don't think we have an exact coding style rules about labels, but i would make `FINISHED` lowercase.

File: core/plannernotes.c
Comment (neolit123) [reply core/plannernotes.c]: `icyl=0; icyl<MAX_CYLINDERS`
please add spaces around `=` and `<`.

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: single tabs instead of double tabs should be used here for the second line of `snprintf(...)`.
same for the rest of the calls bellow.

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: no need to add this empty line.

**Review by neolit123 — CHANGES_REQUESTED** [review]
added a few coding style nitpicks.
thanks.

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: Please see what it looks like now? There is a discrepancy between the rendering of tabs between gedit that I use, and the way that the gitbub editor rends the same contents. See image below in gedit which shows good alignment. I can do some final edits in the github interactive editor. Would this be a way out?
tab-rendering

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: See above

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: See my comment above

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done.

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done. icd_data is passed as pointer.
the first three arguments have now been educed to 2 arguments as suggested.

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: done

**Review by dirkhh — APPROVED** [review]
One tiny style comment, otherwise it looks good.
It's a somewhat bigger commit, but I think it would have been hard to split it up and to make sense of the parts. Well, maybe the "goto finished" part could have been its own commit, but that's really just nit-picking at this point.

**Comment by dirkhh** [comment]
Again, thanks for the patience and persistence, @willemferguson 
And thanks for the detailed reviews, @sfuchs79 @janmulder @janmulder @atdotde

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c:544]: It's a preference (quirk?) that I have. I was reading the code and all the nested if statements were not easy to follow. With the braces the code is easier to understand.

Want me to take it out? Any time.

Comment (dirkhh) [reply core/plannernotes.c:544]: It's not a big deal. We have those everywhere - and since this is a multi line statement, this may indeed make it easier to read. So no, don't worry.

**Comment by dirkhh** [comment]
@neolit123 - are you ok with the changes? I'll wait for you to approve
@atdotde - since this is planner code, do you want to merge it once @neolit123 has approved?

**Comment by atdotde** [comment]
On the phone right now. Will do as soon as I am back to a proper screen and had a glance at the code.

File: core/plannernotes.c
Comment (atdotde) [inline core/plannernotes.c]: Now that I see it, struct dive is no longer needed, is it?

**Review by atdotde — COMMENTED** [review]
Except this one no longer used parameter, it looks fine. Thanks a lot Willem!

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: this `for` loop should have ` { } `.

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c]: Indeed. corrected. Kind regards, willem

**Review by atdotde — APPROVED** [review]

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c]: i would add `{ } ` for this `if` too.

File: core/plannernotes.c
Comment (neolit123) [inline core/plannernotes.c:46]: these look good now.

**Review by neolit123 — CHANGES_REQUESTED** [review]
a couple of small coding style nits remaining. thanks.

File: core/plannernotes.c
Comment (willemferguson) [reply core/plannernotes.c:46]: two last corrections made

**Review by neolit123 — APPROVED** [review]
lgtm, thanks.

**Comment by neolit123** [comment]
there seems to be a stall in the mac build. merging as win32 and linux built fine.

**Pull request merged**
